Annotations extras

PierreGtch opened the pull request:
- closes #12208 
- closes https://github.com/mne-tools/mne-python/pull/12213

<!--

Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://mne.tools/dev/development/contributing.html)
before submitting.

Please be aware that we are a loose team of volunteers so patience is
necessary. Assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Again, thanks for contributing!

-->

#### Reference issue (if any)

This PR replaces  #13213

#### What does this implement/fix?

The difference with #13213 is that the additional information is a `list[dict[str, str | int | float | None]]` instead of a dataframe

PierreGtch commented:
@drammock I implemented here the recommendations in your https://github.com/mne-tools/mne-python/pull/13213#issuecomment-2824466625

agramfort commented on mne/io/egi/tests/test_egi.py:
```diff
@@ -212,7 +212,7 @@ def test_io_egi_mff(events_as_annotations):
         # Grab the first annotation. Should be the first "DIN1" event.
         assert len(raw.annotations)
-        onset, dur, desc, _ = raw.annotations[0].values()
+        onset, dur, desc, _, _ = raw.annotations[0].values()
```
this tells me it's a breaking change but I doubt this will affect many people...

agramfort commented on mne/annotations.py:
```diff
@@ -58,7 +59,48 @@
 
 
-def _check_o_d_s_c(onset, duration, description, ch_names):
+class _DetailsDict(dict):
```
I am not a fan of the word "details". It suggests it's not important. To me it's more extra fields.

agramfort commented on mne/annotations.py:
```diff
@@ -58,7 +59,48 @@
 
-def _check_o_d_s_c(onset, duration, description, ch_names):
+class _DetailsDict(dict):
+    """A dictionary for storing details of annotations.
```
```suggestion
    """A dictionary for storing extra fields of annotations.
```

agramfort commented on mne/annotations.py:
```diff
@@ -146,6 +190,11 @@ class Annotations:
     %(ch_names_annot)s
 
         .. versionadded:: 0.23
+    details : list[dict | None] | None
```
```suggestion
    details : list[dict[str, int | float | str]] | None
```
?

agramfort commented on mne/annotations.py:
```diff
@@ -274,10 +323,12 @@ class Annotations:
 
-    def __init__(self, onset, duration, description, orig_time=None, ch_names=None):
+    def __init__(
+        self, onset, duration, description, orig_time=None, ch_names=None, details=None
```
```suggestion
        self, onset, duration, description, orig_time=None, ch_names=None, extras=None
```
thoughts my native english speaker friends?

agramfort commented on mne/annotations.py:
```diff
@@ -286,6 +337,26 @@ def orig_time(self):
+    def details_data_frame(self):
+        """The details of the Annotations as a DataFrame."""
+        pd = _check_pandas_installed(strict=True)
+        return pd.DataFrame([d if d is not None else {} for d in self.details])
```
I don't get how you can have None in the list

drammock replied:
if you have say 3 annotation spans, and second one is `BAD_EDGE` then that one probably won't have any extra metadata attached to it.  So the list would be `[dict(...), None, dict(...)]`

agramfort replied:
hum... my first reaction is then to have an empty dict unless there is an implicit assumption that the keys in each dictionary should be the same for all annotations unless it's None.

here the design is via an array of structures (list of dicts) while onset, duration and description are structures of arrays (dict-like of arrays). The latter forces all annotations to have the same schema. That would be my inclination to force the same schema unless it's deemed too restrictive.

my 2c

PierreGtch replied:
I modified such that:
- internally, extras is a list of dicts. This simplifies the code and avoids the None case
- when instantiating, the users still have the possibility to pass a `list[dict | None]` for simplicity

agramfort reviewed (commented):
I am onboard with the change although I am not convinced  about the word "details"

PierreGtch commented on mne/annotations.py:
```diff
@@ -436,6 +521,11 @@ def delete(self, idx):
+        if isinstance(idx, int_like):
+            del self.details[idx]
+        elif len(idx) > 0:
+            for i in np.sort(np.arange(len(self.details))[idx])[::-1]:
+                del self.details[i]
```
`idx` can also be a bool array or a slice. I think this is the easiest way to handle all cases

agramfort commented on doc/changes/devel/13228.newfeature.rst:
```diff
@@ -0,0 +1 @@
+Add a ``details`` attribute to :class:`mne.Annotations`, by `Pierre Guetschel`_.
```
```suggestion
Add an ``extras`` attribute to :class:`mne.Annotations`, by `Pierre Guetschel`_.
```

agramfort reviewed (approved):
good from my end. I'll let @drammock do another pass and merge if happy. thx a lot @PierreGtch !

PierreGtch commented:
You're welcome @agramfort :)

@drammock it's ready for another pass

sappelhoff commented:
Edited the PR description to close 

- #12208 
- #12213 

@eioe please let me know if the present PR does not solve the issue and PR that you opened back then, in that case I will revert my edit.

PierreGtch commented:
@drammock the failed test seems unrelated to the PR (Read timed out)

drammock commented on mne/annotations.py:
```diff
@@ -1328,28 +1454,51 @@ def _read_annotations_txt_parse_header(fname):
-        if len(out) == 3:
-            onset, duration, desc = out
+        if columns is None:
+            _check_option("text header", len(out), (3, 4))
+            columns = ["onset", "duration", "description"] + (
+                ["ch_names"] if len(out) == 4 else []
+            )
         else:
-            onset, duration, desc, ch_names = out
+            _check_option(
+                "text header", columns[:3], (["onset", "duration", "description"],)
+            )
+            _check_option("text header len", len(out), (len(columns),))
+        onset, duration, desc = out[:3]
+        i_col = 3
+        if len(columns) > i_col and columns[i_col] == "ch_names":
+            ch_names = out[i_col]
+            i_col += 1
+        if len(columns) > i_col:
+            extras = [
+                {columns[j_col]: out[j_col][i] for j_col in range(i_col, len(columns))}
+                for i in range(len(onset))
+            ]
```
this section here feels a bit hard to maintain. Can you take a few minutes to see if you can simplify the logic or otherwise reduce the complexity?  If you're out of ideas LMK and I can push something.

drammock reviewed (commented):
I think this is close.  I've left several questions below.  One thing I want to make sure of (but ran out of time to check) is that users can modify `extras` in-place (the way they can with `onset`, `duration`, etc).  In other words:

```py
annotations.onset[3] = 7.1  # this works
annotations.extras[3] = dict(reaction_time=0.7)  # this should be possible
annotations.extras[3].update(dict(button="left"))  # this also should be possible
```

PierreGtch commented on mne/annotations.py:
```diff
@@ -58,7 +58,41 @@
 
 
-def _check_o_d_s_c(onset, duration, description, ch_names):
+class _AnnotationsExtrasDict(UserDict):
```
This avoids having to overwrite the `update` method too. This can be tested with `test_annotations::test_extras_dict_raises` that would fail at the last line. 

I can rewrite the `update` method if you think it's best to avoid `UserDict`

PierreGtch commented on mne/annotations.py:
```diff
@@ -350,7 +417,7 @@ def __iter__(self):
             yield self.__getitem__(idx, with_ch_names=with_ch_names)
 
-    def __getitem__(self, key, *, with_ch_names=None):
+    def __getitem__(self, key, *, with_ch_names=None, with_extras=True):
```
This only applies when key is an int and the return type is an `OrderedDict`.
I feel like an extra key in a dict is not a very large change. The cases where this would break something are limited.
At the same time, the backwards-compatible way would force users to call the `__getitem__` method if they want the extras, instead of the more convenient `extras[item]` syntax

PierreGtch commented on mne/annotations.py:
```diff
@@ -286,6 +329,26 @@ def orig_time(self):
+        """The set containing all the keys in all extras dicts."""
+        return {k for d in self.extras for k in d.keys()}
+
+    @property
+    def extras_data_frame(self):
+        """The extras of the Annotations as a DataFrame."""
+        pd = _check_pandas_installed(strict=True)
+        return pd.DataFrame(self.extras)
```
Ok, I will remove this property and add a comment in the `extra` property docstring indicating how they can be converted to a dataframe

PierreGtch commented on mne/annotations.py:
```diff
@@ -1100,10 +1207,21 @@ def _write_annotations(fid, annotations):
 
 def _write_annotations_csv(fname, annot):
+    if len(annot.extras_columns) > 0:
+        warn(
+            "Reading extra annotation fields from CSV is not supported. "
+            "The extra fields will be written but not loaded when reading."
+        )
```
I removed it by trying to automatically cast the strings in CSV or TXT files into `int`, `float`, `str` or `None`. But it feels a bit sketchy... what do you think?

There are very little assumptions. For example, the same key my have different types on different rows. Also, an empty string could be `None`, `np.nan` or `""`

drammock replied:
ah, that makes sense.  A couple of thoughts:

- If Pandas were present on the system, we could read from CSV using it's "infer dtypes" capabilities. So maybe check for pandas in the reader function (`_soft_import("pandas", strict=False)`), and if it's not present, warn that everything will be read in as strings?
- Perhaps it would also be worth doing a check when *writing*, to see if the dtypes in a given column are consistent.  If they are not, perhaps a raise warning like 
  > annotations.extra field "foo" has heterogeneous dtypes (string, float); some dtypes may be set incorrectly when loading the data from CSV

  and/or

  > annotations.extra field "foo" has heterogeneous dtypes (string, None); None values will be written to CSV as ""

drammock replied:
WDYT about that @PierreGtch ?

PierreGtch replied:
I like your solution. simpler to document than a custom function. I will implement both checks when writing

PierreGtch commented:
@drammock I addressed your comments but left some questions. In particular:

> ```python
> annotations.extras[3].update(dict(button="left"))  # this also should be possible
> ```
This is possible and the type of the new extras will be checked

> ```python
> annotations.extras[3] = dict(reaction_time=0.7)  # this should be possible
> ```
This is also possible but the type of the new extras will NOT be checked. You think this would be an issue?
To enforce it, a possibility would be to also write a custom list class which would overwrite `__setitem__`  to replace the input `dict` with an `_AnnotationsExtraDict` 

Feel free to merge if you are satisfied with the current state :)

drammock commented:
> This is also possible but the type of the new extras will NOT be checked. You think this would be an issue?

I think this may be the more common use case (it more closely mirrors what users currently do with the `description` field for example).  A custom subclass of `list` is indeed probably the right way to solve it; I'm doing something very similar here:

https://github.com/mne-tools/mne-python/pull/13059/files#diff-8958de0220265a954fac8d4e75521b1917748cdd5e15ce759b5e627b1d401b7cR779-R783

PierreGtch commented:
All good @drammock :)

drammock commented on mne/annotations.py:
```diff
@@ -1368,6 +1443,13 @@ def read_annotations(
 
+    The dtypes of the extra fields will automatically be inferred
+    by pandas. If some fields have heterogeneous types on the
+    different rows, this automatic inference may return unexpecterd
```
```suggestion
    different rows, this automatic inference may return unexpected
```

drammock reviewed (approved):
Thanks for sticking with this @PierreGtch. I'm pretty happy with where this ended up. Approving (modulo one docstring typo) but would like to wait on merging until at least one other maintainer has had a look (mostly because I've only read the code, haven't actually had time to play around with the feature interactively, and tomorrow I'm off to a conference for a week so I can't do it soon)

PierreGtch commented:
thanks @drammock for your time on this PR!

larsoner commented on mne/annotations.py:
```diff
@@ -58,7 +58,103 @@
+    """
+
+    def __setitem__(self, key: str, value: str | int | float | None) -> None:
+        _validate_type(key, str, "key", "string")
```
This makes our behavior in this one function different from elsewhere in MNE-Python
```suggestion
        _validate_type(key, str, "key")
```
If we really want to be more verbose, we should update `_validate_type` to deal with `str` in the list of acceptable types to write it as `"string"` rather than `"str"` so that all places in MNE-Python benefit from it (and consistency is maintained).

But I'm not sure we should... following this logic one could argue you should say "integer" instead of "int", "floating point value" instead of "float", etc...

PierreGtch replied:
I see, thanks for the lecture :D

larsoner commented on mne/annotations.py:
```diff
@@ -58,7 +58,103 @@
+        _validate_type(key, str, "key", "string")
+        if key in ("onset", "duration", "description", "ch_names"):
+            raise ValueError(f"Key '{key}' is reserved and cannot be used in extras.")
+        _validate_type(
+            value, (str, int, float, None), "value", "string, int, float or None"
+        )
```
```suggestion
        _validate_type(
            value, (str, int, float, None), "value",
        )
```

larsoner commented on mne/annotations.py:
```diff
@@ -274,10 +377,12 @@ class Annotations:
 
-    def __init__(self, onset, duration, description, orig_time=None, ch_names=None):
+    def __init__(
+        self, onset, duration, description, orig_time=None, ch_names=None, extras=None
```
```suggestion
        self, onset, duration, description, orig_time=None, ch_names=None, *, extras=None
```

larsoner commented on mne/annotations.py:
```diff
@@ -286,6 +391,25 @@ def orig_time(self):
+    @property
+    def _extras_columns(self) -> set[str]:
+        """The set containing all the keys in all extras dicts."""
+        return {k for d in self.extras for k in d.keys()}
```
```suggestion
        return {k for d in self.extras for k in d}
```

larsoner commented on mne/annotations.py:
```diff
@@ -372,10 +503,11 @@ def __getitem__(self, key, *, with_ch_names=None):
 
     @fill_doc
-    def append(self, onset, duration, description, ch_names=None):
+    def append(self, onset, duration, description, ch_names=None, extras=None):
```
```suggestion
    def append(self, onset, duration, description, ch_names=None, *, extras=None):
```

larsoner commented on mne/annotations.py:
```diff
@@ -822,9 +971,17 @@ def set_annotations(self, annotations, on_missing="raise", *, verbose=None):
         return self
 
-    def get_annotations_per_epoch(self):
+    def get_annotations_per_epoch(self, with_extras=False):
```
I think `get_annotations_per_epch(True)` isn't verbose enough, so let's do
```suggestion
    def get_annotations_per_epoch(self, *, with_extras=False):
```

larsoner commented on mne/annotations.py:
```diff
@@ -893,11 +1050,13 @@ def get_annotations_per_epoch(self):
         return epoch_annot_list
 
-    def add_annotations_to_metadata(self, overwrite=False):
+    def add_annotations_to_metadata(self, overwrite=False, with_extras=True):
```
```suggestion
    def add_annotations_to_metadata(self, overwrite=False, *, with_extras=True):
```

larsoner reviewed (commented):
Looks good! Just minor stuff from me really. Can you push the next commit with `[circle full]` in the commit message? That will make CircleCI build all examples, which I think would be a nice sanity check (and save us some possible post-merge pain in case we missed something)

drammock commented:
105 CircleCI failures (!).  Luckily they're all of 2 types:

1. TypeError: Annotations.append() takes from 4 to 5 positional arguments but 6 were given

<details>
examples/datasets/kernel_phantom.py
examples/datasets/opm_data.py
examples/decoding/decoding_spatio_temporal_source.py
examples/decoding/decoding_time_generalization_conditions.py
examples/decoding/decoding_unsupervised_spatial_filter.py
examples/decoding/decoding_xdawn_eeg.py
examples/decoding/ems_filtering.py
examples/decoding/linear_model_patterns.py
examples/inverse/compute_mne_inverse_epochs_in_label.py
examples/inverse/compute_mne_inverse_raw_in_label.py
examples/inverse/dics_epochs.py
examples/inverse/dics_source_power.py
examples/inverse/evoked_ers_source_power.py
examples/inverse/mne_cov_power.py
examples/inverse/multi_dipole_model.py
examples/inverse/multidict_reweighted_tfmxne.py
examples/inverse/psf_ctf_vertices_lcmv.py
examples/io/elekta_epochs.py
examples/preprocessing/css.py
examples/preprocessing/define_target_events.py
examples/preprocessing/eeg_csd.py
examples/preprocessing/eog_artifact_histogram.py
examples/preprocessing/eog_regression.py
examples/preprocessing/find_ref_artifacts.py
examples/preprocessing/ica_comparison.py
examples/preprocessing/movement_compensation.py
examples/preprocessing/movement_detection.py
examples/preprocessing/muscle_ica.py
examples/preprocessing/otp.py
examples/preprocessing/xdawn_denoising.py
examples/simulation/simulate_evoked_data.py
examples/simulation/simulate_raw_data.py
examples/stats/cluster_stats_evoked.py
examples/stats/fdr_stats_evoked.py
examples/stats/linear_regression_raw.py
examples/stats/sensor_permutation_test.py
examples/time_frequency/compute_csd.py
examples/time_frequency/compute_source_psd_epochs.py
examples/time_frequency/source_label_time_frequency.py
examples/time_frequency/source_power_spectrum_opm.py
examples/time_frequency/source_power_spectrum.py
examples/time_frequency/source_space_time_frequency.py
examples/time_frequency/temporal_whitening.py
examples/time_frequency/time_frequency_global_field_power.py
examples/visualization/3d_to_2d.py
examples/visualization/channel_epochs_image.py
examples/visualization/eeg_on_scalp.py
examples/visualization/evoked_whitening.py
examples/visualization/eyetracking_plot_heatmap.py
examples/visualization/meg_sensors.py
examples/visualization/topo_compare_conditions.py
examples/visualization/topo_customized.py
tutorials/epochs/10_epochs_overview.py
tutorials/epochs/15_baseline_regression.py
tutorials/epochs/20_visualize_epochs.py
tutorials/epochs/50_epochs_to_data_frame.py
tutorials/epochs/60_make_fixed_length_epochs.py
tutorials/evoked/10_evoked_overview.py
tutorials/evoked/30_eeg_erp.py
tutorials/evoked/40_whitened.py
tutorials/forward/20_source_alignment.py
tutorials/forward/90_compute_covariance.py
tutorials/intro/10_overview.py
tutorials/intro/15_inplace.py
tutorials/intro/20_events_from_raw.py
tutorials/intro/30_info.py
tutorials/intro/40_sensor_locations.py
tutorials/intro/70_report.py
tutorials/inverse/30_mne_dspm_loreta.py
tutorials/inverse/50_beamformer_lcmv.py
tutorials/inverse/70_eeg_mri_coords.py
tutorials/inverse/80_brainstorm_phantom_elekta.py
tutorials/inverse/95_phantom_KIT.py
tutorials/io/70_reading_eyetracking_data.py
tutorials/machine-learning/50_decoding.py
tutorials/preprocessing/10_preprocessing_overview.py
tutorials/preprocessing/15_handling_bad_channels.py
tutorials/preprocessing/20_rejecting_bad_data.py
tutorials/preprocessing/30_filtering_resampling.py
tutorials/preprocessing/35_artifact_correction_regression.py
tutorials/preprocessing/40_artifact_correction_ica.py
tutorials/preprocessing/45_projectors_background.py
tutorials/preprocessing/50_artifact_correction_ssp.py
tutorials/preprocessing/55_setting_eeg_reference.py
tutorials/preprocessing/59_head_positions.py
tutorials/preprocessing/60_maxwell_filtering_sss.py
tutorials/preprocessing/90_eyetracking_data.py
tutorials/raw/10_raw_overview.py
tutorials/raw/20_event_arrays.py
tutorials/raw/30_annotate_raw.py
tutorials/raw/40_visualize_raw.py
tutorials/simulation/70_point_spread.py
tutorials/simulation/80_dics.py
tutorials/stats-sensor-space/40_cluster_1samp_time_freq.py
tutorials/stats-sensor-space/50_cluster_between_time_freq.py
tutorials/stats-sensor-space/70_cluster_rmANOVA_time_freq.py
tutorials/stats-sensor-space/75_cluster_ftest_spatiotemporal.py
tutorials/stats-source-space/20_cluster_1samp_spatiotemporal.py
tutorials/stats-source-space/60_cluster_rmANOVA_spatiotemporal.py
tutorials/time-freq/10_spectrum_class.py
tutorials/time-freq/20_sensors_time_frequency.py
tutorials/visualization/10_publication_figure.py
</details>

2. TypeError: Annotations.__init__() takes from 4 to 6 positional arguments but 7 were given

<details>
examples/preprocessing/epochs_metadata.py
tutorials/clinical/20_seeg.py
tutorials/epochs/40_autogenerate_metadata.py
</details>

drammock commented:
> @larsoner thanks for the review! Is this the kind of test you had in mind?

Looking at what you had to change in 740694804c0b158def50f07ef838709da6b84c2f, I think tests that called `__iadd__`, `_read_annotations_csv`, and `_read_annotations_fif` (either directly, or implicitly via a public API call) should have caught the same failures that running the tutorials caught. My guess is that is what @larsoner meant.

PierreGtch commented:
@drammock @larsoner The issue was introduced in the last-minute suggestion from code review: 9e67e3b
And the non-circle-full tests did catch it. Not caught in the initial commit 9e67e3b  because tests were aborted due to pre-commit, but in 3621595 we see that ubuntu-pip fails with 928 errors.
So it should be all good to merge
